0000-zmix/README.md

```diff
@@ -0,0 +1,395 @@
+
+In the example above, there are three components: 
+
+- What do we want to prove? That there exists an `x` such that `A = g^x`
+and `B = h^x`.
+
+- How do we know this is true? Because we know `x`.
+
+- How can we prove it? With a zero-knowledge proof that the prover produces
+and the verifier verifies.
+
+In zmix, we have data types corresponding to these three components:
+
+- `ProofSpec`: specifies what we want to prove
+
+- `Witness`: contains the secrets that satisfy the `ProofSpec`.
+
+- `Proof`: the zero-knowledge proof. 
+
+The library provides two features: construction and verification of
+zero-knowledge proofs. 
+
+To construct a proof, zmix requires a `ProofSpec s` and a `Witness w`
+(which is a valid witness for `p`), and will output a `Proof p`. 
+
+To verify a proof, zmix requires a `ProofSpec s` and a `Proof p`, and will
+output a boolean indicating whether the proof was valid (true) or invalid
+(false).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Conceptually, the zmix library will offer the functions:
+
+* `generate_proof(s: ProofSpec, w: Witness) -> Result<Proof, Error>`
+* `verify_proof(p: Proof, s: ProofSpec) -> Result<(), Error>`
+
+where the proof specification contains
+
+1. the number of secrets involved in the zero-knowlege proof (we call these
+_messages_), and
+1. a list of _statements_, where each statement represents a _part_ of the overall zero-knowlege proof:
+
+```
+pub struct ProofSpec {
+    message_count: usize,
+    statements: Vec<Statement>,
+    ...
+}
+```
+The library will offer and support various types of statements, such as signatures and commitments.
+To combine and connect all the statements (i.e., the proof's parts) contained in the proof specification, that is, glue all parts together into a single zero-knowledge proof, the library will use _Schnorr proofs_.
+
+For generating a Schnorr proof that encompasses multiple statements, for each statement one requires a _hash contribution_ on the one hand, and a _proof contribution_ based on an aggregation of the hash contributions on the other hand.
+In particular, the library will implement a proof orchestrator that conceptually generates Schnorr proofs as follows:
+
+![Generation flow](zmix_proof_generation.png)
+
+For verifying such a Schnorr proof, one again requires (re-computed) hash contributions for each statement.
+In particular, the library's proof orchestrator will conceptually verify a Schnorr proof as follows:
+
+![Verification flow](zmix_proof_verification.png)
+
+## Proof Modules
+
+In accordance with the sequence diagrams shown previously, to (re-)generate the hash contributions and the proof contributions, the library will implement the following _proof module_ interface for each supported statement type:
+
+```
+pub trait ProofModule {
+    fn get_hash_contribution(
+        statement: Statement,
+        witness: StatementWitness,
+        message_r_values: Vec<Vec<u8>>,
```

Comment:
Assuming `message_r_values` are the blinding factors, why not rename `message_r_values` to something like that?